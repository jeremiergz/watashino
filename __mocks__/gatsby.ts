import Gatsby from 'gatsby';

module.exports = {
  ...Gatsby,
  graphql: jest.fn(),
  navigate: jest.fn(),
  StaticQuery: jest.fn(),
  useStaticQuery: jest.fn(),
};
