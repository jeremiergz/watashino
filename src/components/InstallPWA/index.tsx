import React, { useEffect } from 'react';

type BeforeInstallPromptEvent = Event & {
  prompt(): Promise<void>;
};

const eventName = 'beforeinstallprompt';

const InstallPWA: React.FC = () => {
  useEffect(() => {
    const handler = (e: BeforeInstallPromptEvent) => {
      e.preventDefault();
      e.prompt();
    };
    window.addEventListener(eventName, handler);
    return () => {
      window.removeEventListener(eventName, handler);
    };
  }, []);
  return null;
};

InstallPWA.displayName = 'InstallPWA';

export default InstallPWA;
