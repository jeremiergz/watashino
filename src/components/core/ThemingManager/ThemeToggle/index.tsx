import React from 'react';
import { useTheming } from '..';
import Button from '../../../common/Button';
import Flex from '../../../primitives/Flex';
import Text from '../../../primitives/Text';
import Moon from '../../../svg/Moon';
import Sun from '../../../svg/Sun';

const iconMapping = {
  dark: Moon,
  light: Sun,
};

const ThemeToggle = () => {
  const { theme, toggle } = useTheming();
  const ThemeIcon = iconMapping[theme.type];
  return (
    <Button onClick={() => toggle()} position="absolute" right={3} top={3} variant="outlined" width={72}>
      <Flex alignItems="center" justifyContent="space-between" width="100%">
        <ThemeIcon height={20} width={20} />
        <Text color="gray" fontFamily={theme.fonts.main} fontSize={14}>
          {theme.type}
        </Text>
      </Flex>
    </Button>
  );
};

ThemeToggle.displayName = 'ThemeToggle';

export default ThemeToggle;
