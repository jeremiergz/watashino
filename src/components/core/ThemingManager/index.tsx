import React, { createContext, PropsWithChildren, useContext, useState } from 'react';
import { ThemeProvider } from 'styled-components';
import { DarkTheme, LightTheme, Theme } from '../../../theme';

type ThemingManagerProps = PropsWithChildren<{}>;

const defaultMode = 'light';
const localStorageKey = 'theme.mode';

const ThemingManagerContext = createContext<{ mode: ThemeMode; theme: Theme; toggle: () => void }>({
  mode: defaultMode,
  theme: LightTheme,
  toggle: () => null,
});

function useTheming() {
  return useContext(ThemingManagerContext);
}

const ThemingManager = ({ children }: ThemingManagerProps) => {
  const isBrowser = typeof window !== 'undefined';
  const storedMode = (isBrowser && (localStorage.getItem(localStorageKey) as ThemeMode)) || defaultMode;
  const [mode, setMode] = useState<ThemeMode>(storedMode);
  const toggle = () => {
    const newMode = mode === 'light' ? 'dark' : 'light';
    setMode(newMode);
    if (isBrowser) localStorage.setItem(localStorageKey, newMode);
  };
  const theme = mode === 'light' ? LightTheme : DarkTheme;
  return (
    <ThemingManagerContext.Provider value={{ mode, theme, toggle }}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ThemingManagerContext.Provider>
  );
};

ThemingManager.displayName = 'ThemingManager';

export { useTheming };
export default ThemingManager;
