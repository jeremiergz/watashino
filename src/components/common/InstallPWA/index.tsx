import { graphql, useStaticQuery } from 'gatsby';
import Img from 'gatsby-image';
import { trackCustomEvent } from 'gatsby-plugin-google-analytics';
import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import Flex from '../../primitives/Flex';
import Text from '../../primitives/Text';
import A2HSIcon from '../../svg/A2HS';
import CloseIcon from '../../svg/Close';
import Button from '../Button';

type BeforeInstallPromptEvent = Event & {
  prompt(): Promise<void>;
};

const Container = styled(Flex)`
  align-items: center;
  bottom: 0;
  left: 0;
  margin: auto;
  position: fixed;
  right: 0;
  transform: ${({ isDisplayed }: { isDisplayed: boolean }) =>
    isDisplayed ? 'translate3d(0, 0, 0)' : 'translate3d(0, 100%, 0)'};
  transition: transform 250ms;
`;

const eventName = 'beforeinstallprompt';

const InstallPWA = () => {
  const { appIcon } = useStaticQuery<GraphQL.InstallPwaDataQuery>(graphql`
    query InstallPWAData {
      appIcon: file(relativePath: { eq: "profile-pic.png" }) {
        childImageSharp {
          fixed(height: 32, width: 32) {
            ...GatsbyImageSharpFixed_withWebp
          }
        }
      }
    }
  `);
  const [isDisplayed, setIsDisplayed] = useState(true);
  const [isPWAReady, setIsPWAReady] = useState(false);
  const [promptInstall, setPromptInstall] = useState<BeforeInstallPromptEvent>(null);
  const handleClose = () => {
    trackCustomEvent({
      action: 'click',
      category: 'Install PWA App',
      value: 0,
    });
    setIsDisplayed(false);
  };
  const handleInstall = () => {
    trackCustomEvent({
      action: 'click',
      category: 'Install PWA App',
      value: 1,
    });
    if (promptInstall) promptInstall.prompt();
  };
  useEffect(() => {
    const handler = (e: BeforeInstallPromptEvent) => {
      e.preventDefault();
      setIsPWAReady(true);
      setPromptInstall(e);
    };
    window.addEventListener(eventName, handler);
    return () => window.removeEventListener(eventName, handler);
  }, []);
  return (
    <Container
      backgroundColor="secondary"
      borderTopLeftRadius={{ mobileL: 8 }}
      borderTopRightRadius={{ mobileL: 8 }}
      isDisplayed={isPWAReady && isDisplayed}
      justifyContent={{ _: 'center', mobileL: 'flex-start' }}
      padding={2}
      position="relative"
      width={{ mobileL: 256 }}
    >
      <Flex alignItems="center" justifyContent="center">
        <Button onClick={handleInstall} paddingX={3}>
          <Img fixed={appIcon.childImageSharp.fixed} imgStyle={{ height: 32, width: 32 }} />
          <Text fontSize={18} marginLeft={3} marginRight={2}>
            Install App
          </Text>
          <A2HSIcon fill="white" height={24} width={24} />
        </Button>
      </Flex>
      <Button onClick={handleClose} paddingRight={3} position="absolute" right={0}>
        <CloseIcon fill="white" />
      </Button>
    </Container>
  );
};

export default InstallPWA;
