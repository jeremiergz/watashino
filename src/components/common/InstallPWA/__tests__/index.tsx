import * as Gatsby from 'gatsby';
import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import InstallPWA from '..';
import ThemingManager from '../../../core/ThemingManager';

describe('components/common/InstallPWA Test Suite', () => {
  beforeAll(() => {
    const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery');
    useStaticQuery.mockImplementation(() => ({
      appIcon: {
        childImageSharp: {
          fixed: {
            base64: 'data:image/png;base64',
            aspectRatio: 1,
            src: '',
            srcSet: '',
          },
        },
      },
    }));
  });

  it('matches snapshot', () => {
    const jsx = (
      <ThemingManager>
        <InstallPWA />
      </ThemingManager>
    );
    expect(ReactTestRenderer.create(jsx).toJSON()).toMatchSnapshot();
  });
});
