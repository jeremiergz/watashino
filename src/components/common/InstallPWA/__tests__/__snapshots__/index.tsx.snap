// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/InstallPWA Test Suite matches snapshot 1`] = `
<div
  className="sc-AykKC jqflLh sc-AykKG gPESru"
  display="flex"
  width={
    Object {
      "mobileL": 256,
    }
  }
>
  <div
    className="sc-AykKC cHQUle"
    display="flex"
  >
    <button
      className="sc-AykKF jXEPwR"
      onClick={[Function]}
    >
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "display": "inline-block",
            "height": undefined,
            "overflow": "hidden",
            "position": "relative",
            "width": undefined,
          }
        }
      >
        <img
          alt=""
          aria-hidden={true}
          src="data:image/png;base64"
          style={
            Object {
              "height": 32,
              "left": 0,
              "objectFit": "cover",
              "objectPosition": "center",
              "opacity": 1,
              "position": "absolute",
              "top": 0,
              "transitionDelay": "500ms",
              "width": 32,
            }
          }
        />
        <picture>
          <source
            srcSet=""
          />
          <img
            alt=""
            loading="lazy"
            onLoad={[Function]}
            src=""
            srcSet=""
            style={
              Object {
                "height": 32,
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 500ms",
                "width": 32,
              }
            }
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcset=\\"\\" /><img loading=\\"lazy\\" src=\\"\\" alt=\\"\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
      <span
        className="sc-AykKD jjaliH"
        fontSize={18}
      >
        Install App
      </span>
      <svg
        className="sc-AykKE dOqYrG"
        height={24}
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M18 1.01L8 1c-1.1 0-2 .9-2 2v3h2V5h10v14H8v-1H6v3c0 1.1.9 2 2 2h10c1.1 0 2-.9 2-2V3c0-1.1-.9-1.99-2-1.99zM10 15h2V8H5v2h3.59L3 15.59 4.41 17 10 11.41z"
          fill="white"
        />
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
      </svg>
    </button>
  </div>
  <button
    className="sc-AykKF iAjbny"
    onClick={[Function]}
  >
    <svg
      className="sc-AykKE QaHlM"
      height={32}
      viewBox="0 0 24 24"
      width={32}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        fill="white"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </button>
</div>
`;
